### PR TITLE
Setup a CI job for Go code linting. Fixes # 3576 

### LINKED ISSUE #3576: Setup a CI job for Go code linting
We would benefit from having a CI job that runs Go linters to detect unused code, unchecked errors, ...

FWIW the Prometheus project uses [golangci-lint](https://github.com/golangci/golangci-lint).

### PR DESCRIPTION
**Pull request by Amab**
- Added a GitHub Action for Golang linter using the following instructions [golangci-lint-action (How to use)](https://github.com/golangci/golangci-lint-action#how-to-use).
- Added the `.golangci.yml` file.
- Added the `make check-golang` command to run the linter locally.

Fixes #3576

### CODE REVIEW

**Review by paulfantom — APPROVED** [review]
LGTM

I think we can merge it and fix lint issues in next PRs as check is not blocking now. After issues are fixed we should probably mark this as a "required" check. @prometheus-operator/prometheus-operator-reviewers wdyt?

**Comment by simonpasquier** [comment]
Can we add a `link` target to the Makefile too? So that you can run the linter before pushing to GitHub.

We can also include a `.golangci.yml` file that would only enable specific linters to begin with (example from [Prometheus](https://github.com/prometheus/prometheus/blob/master/.golangci.yml)). Enabling only the `unused` linter could be done in this PR because it wouldn't require too many changes. Then following PRs can gradually extend the scope.

**Comment by simonpasquier** [comment]
:wave: @Amab are you still ok to work on it?

**Comment by Amab** [comment]
@simonpasquier yes, I have already applied the requested changes.

File: .golangci.yml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+run:
+  modules-download-mode: vendor
```
Comment (simonpasquier) [inline .golangci.yml:2]:
```suggestion
```

**Review by simonpasquier — COMMENTED** [review]
We don't vendor dependencies anymore so you need to remove any `vendor` directory from this PR.

File: Makefile [hunk 2]
```diff
@@ -32,6 +32,7 @@ JSONNET_BINARY=$(TOOLS_BIN_DIR)/jsonnet
 JSONNETFMT_BINARY=$(TOOLS_BIN_DIR)/jsonnetfmt
 SHELLCHECK_BINARY=$(TOOLS_BIN_DIR)/shellcheck
 PROMLINTER_BINARY=$(TOOLS_BIN_DIR)/promlinter
+GOLANGCILINTER_BINARY=$(TOOLS_BIN_DIR)/golangci-lint
 TOOLING=$(PO_DOCGEN_BINARY) $(CONTROLLER_GEN_BINARY) $(EMBEDMD_BINARY) $(GOBINDATA_BINARY) $(JB_BINARY) $(GOJSONTOYAML_BINARY) $(JSONNET_BINARY) $(JSONNETFMT_BINARY) $(SHELLCHECK_BINARY) $(PROMLINTER_BINARY)
```
[the remark this answers is not in the document]
Comment (Amab) [reply Makefile]: Fixed.

File: .golangci.yml [hunk 3]
```diff
@@ -0,0 +1,7 @@
...
+  deadline: 10m
+
+linters:
+  enable:
+    - unused
```
[the remark this answers is not in the document]
Comment (Amab) [reply .golangci.yml:7]: Fixed.

File: .golangci.yml (hunk 1 above)
Comment (Amab) [reply .golangci.yml]: Fixed.

**Comment by simonpasquier** [comment]
Can you fix the unused errors?

**Comment by Amab** [comment]
> Can you fix the unused errors?

Done.

**Comment by simonpasquier** [comment]
Thanks! lgtm for me and can be merged on green.

**Pull request merged**